src/s390_ecc.c

```diff
@@ -598,12 +602,96 @@ ECDSA_SIGN_REPLY* make_ecdsa_sign_request(const ICA_EC_KEY *privkey,
 	offset = make_cprbx((struct CPRBX *)cbrbmem, parmblock_len, preqcblk, prepcblk);
 	offset += make_ecdsa_sign_parmblock((ECDSA_PARMBLOCK_PART1*)(cbrbmem+offset), hash, hash_length);
 	offset += make_keyblock_length((ECC_KEYBLOCK_LENGTH*)(cbrbmem+offset), keyblock_len);
-	offset += make_ecdsa_private_key_token(cbrbmem+offset, privkey, curve_type);
+	offset += make_ecdsa_private_key_token(cbrbmem+offset, privkey, X, Y, curve_type);
 	finalize_xcrb(xcrb, preqcblk, prepcblk);
 
     return (ECDSA_SIGN_REPLY*)prepcblk;
 }
 
+/**
+ * calculate the public (X,Y) values for the given private key, if necessary.
+ */
+unsigned int provide_pubkey(const ICA_EC_KEY *privkey, unsigned char *X, unsigned char *Y)
+{
+	EC_KEY *eckey = NULL;
+	EC_POINT *pub_key = NULL;
+	const EC_GROUP *group = NULL;
+	BIGNUM *bn_d = NULL, *bn_x = NULL, *bn_y = NULL;
+	char* x_str = NULL;
+	char* y_str = NULL;
+	int privlen = -1;
+	unsigned int i, n, rc;
+
+	if (privkey == NULL || X == NULL || Y == NULL) {
+		return EFAULT;
+	}
+
+	privlen = privlen_from_nid(privkey->nid);
+	if (privlen < 0) {
+		return EFAULT;
+	}
+
+	/* Check if (X,Y) already available */
+	if (privkey->X != NULL && privkey->Y != NULL) {
+		memcpy(X, privkey->X, privlen);
+		memcpy(Y, privkey->Y, privlen);
+		return 0;
+	}
+
+	/* Get (D) as BIGNUM */
+	if ((bn_d = BN_bin2bn(privkey->D, privlen, NULL)) == NULL) {
+		return EFAULT;
+	}
+
+	/* Calculate public (X,Y) values */
+	eckey = EC_KEY_new_by_curve_name(privkey->nid);
+	EC_KEY_set_private_key(eckey, bn_d);
+	group = EC_KEY_get0_group(eckey);
+	pub_key = EC_POINT_new(group);
+	if (!EC_POINT_mul(group, pub_key, bn_d, NULL, NULL, NULL)) {
+		rc = EFAULT;
+		goto end;
+	}
+
+	/* Get (X,Y) as BIGNUMs */
+	bn_x = BN_new();
+	bn_y = BN_new();
+	if (!EC_POINT_get_affine_coordinates_GFp(group, pub_key, bn_x, bn_y, NULL)) {
+		rc = EFAULT;
+		goto end;
+	}
+
+	/* Format (X) as char array, with leading zeros if necessary */
+	x_str = BN_bn2hex(bn_x);
+	n = privlen - strlen(x_str) / 2;
```

Comment:
is this guaranteed to be divisible by two? if not: does this still work ?

Reply:
Yes .... for example, when using the p521 curve, x_str sometimes has only 65 bytes (Could be even less, depending on the actual value of bn_x). Dividing by 2 results in a rounded int of 32 bytes. However .... I think this division is not necessary at all. What I want to do here, is pre-pending 0x00 bytes so that the resulting char array for X has exactly 66 bytes in this example. The code works, but does too much. It would work also without dividing by 2:  

x_str = BN_bn2hex(bn_x);    // <- e.g. 65 bytes
n = privlen - strlen(x_str);    // <- n = 66 - 65 = 1
for (i = 0; i < n; i++)           
	X[i] = 0x00;                  // <- X[0] = 0x00
BN_bn2bin(bn_x, &(X[n]));   // <- place the bn_x value at X[1] 
  
What do you think?

Reply:
No, the division by 2 was needed because 1 byte is encoded as 2 hex digits. Question is why do you encode the binary value to a hex *string*, measure the length of that string and divide it by two to get the number of bytes instead of using BN_num_bytes(bn_x)? i didnt get it.

Reply:
Agreed. BN_num_bytes() will do it.